### PR TITLE
Running C-Lisp code on Nvidia GPUs

### PR DESCRIPTION
**Pull request by GlowingScrewdriver**
This PR introduces a small framework for compiling and launching kernels written in C-Lisp on an Nvidia GPU

Documentation on this issue can be found here:
https://outline.von-neumann.ai/s/d0cd5eb9-2e15-4fa4-bd17-c3911f305008/doc/attacking-ptx-kKZJzn5Uem

### CODE REVIEW

**Comment by GlowingScrewdriver** [comment]
Address space markers have been implemented in C-Lisp. Test kernels [`vecadd.sexp`](src/backend/tests/kernelprog/vecadd/vecadd.sexp) and [`matmul.sexp`](src/backend/tests/kernelprog/matmul/matmul.sexp) now use address space 1 for kernel parameters.

File: src/backend/brilisp.py [hunk 1]
```diff
@@ -26,7 +26,10 @@ def gen_function(expr):
 
 def gen_type(typ):
     if is_list(typ):
-        return {typ[0]: gen_type(typ[1])}
+        retval = {typ[0]: gen_type(typ[1])}
+        if len(typ) > 2 and is_list(typ[2]):
```
Comment (chsasank) [inline src/backend/brilisp.py:30]: assert that `typ[0]` is ptr is it's a list. also check that typ[2][1] is addrspace.

File: src/backend/llvm.py [hunk 2]
```diff
@@ -46,7 +46,10 @@ def generate(self, bril_prog):
     def gen_type(self, type):
         if isinstance(type, dict):
             if "ptr" in type:
-                return self.gen_type(type["ptr"]).as_pointer()
+                type_obj = self.gen_type(type["ptr"]).as_pointer()
```
Comment (chsasank) [inline src/backend/llvm.py:49]: as_pointer take addrspace as argument

**Comment by chsasank** [comment]
Do we also need addrspace for alloc? It looks like so:

from [docs](https://llvm.org/docs/LangRef.html#alloca-instruction):

> [‘alloca’ Instruction](https://llvm.org/docs/LangRef.html#id2103)
> Syntax: <result> = alloca [inalloca] <type> [, <ty> <NumElements>] [, align <alignment>] [, addrspace(<num>)]     ; yields type addrspace(num)*:result
>  The ‘alloca’ instruction allocates memory on the stack frame of the currently executing function, to be automatically released when this function returns to its caller. If the address space is not explicitly specified, the object is allocated in the alloca address space from the [datalayout string](https://llvm.org/docs/LangRef.html#langref-datalayout).

But llvmlite doesn't seem to support it. May be we don't need to implement at this point of time.

File: src/backend/llvm.py (hunk 2 above)
Comment (GlowingScrewdriver) [reply src/backend/llvm.py:49]: I thought it's cleaner this way, because there is only a single call to gen_type. Should I change it?

**Comment by GlowingScrewdriver** [comment]
> Do we also need addrspace for alloc? It looks like so:

I don't think we need it as of now. Using `alloca` in the kernel without specifying `addrspace` works, and yet doesn't cause `llc` to insert address space casts.

On the other hand, I got an error from the CUDA driver API when I specified `addrspace` on a stack allocation. This is what I tried:
* Generate `vecadd.ll` using `make vecadd.ll`
* Add `addrspace(1)` to one of the `alloca` instructions,
  e.g. `%"tmp_clisp-ckravzof" = alloca i32`
  -> `%"tmp_clisp-ckravzof" = alloca i32, addrspace(1)`
* Correspondingly tweak the instructions that access this pointer
  e.g. `store i32 %"tmp_clisp-ckravzof.1", i32* %"tmp_clisp-ckravzof"`
  -> `store i32 %"tmp_clisp-ckravzof.1", i32 addrspace(1)* %"tmp_clisp-ckravzof"`
* Run the tweaked kernel using `make vecadd.run`

The error reported by the driver: `cuCtxSynchronize() returned non-zero status 717`. Result code 717 (`CUDA_ERROR_INVALID_ADDRESS_SPACE` in CUDA terms) indicates that an instruction was used on data in the wrong address space, as per [Nvidia documentation](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1gc6c391505e117393cc2558fff6bfc2e9)

File: src/backend/brilisp.py (hunk 1 above)
Comment (GlowingScrewdriver) [reply src/backend/brilisp.py:31]: Done

File: src/backend/llvm.py (hunk 2 above)
Comment (GlowingScrewdriver) [reply src/backend/llvm.py]: Done

**Pull request merged**